### ISSUE TITLE
_drush_backend_get_stdin() return FALSE on Windows 10

### ISSUE THREAD

**Issue by xandeadx** [drush-ops/drush#3930]
I work on issue #3315 and found a bug. If execute command `drush make ...` drush creates new process using [proc_open()](https://github.com/drush-ops/drush/blob/8.x/includes/backend.inc#L379).
Step 2 - drush send options in process stdin using [fwrite()](https://github.com/drush-ops/drush/blob/8.x/includes/backend.inc#L382).
Step 3 - in the new process drush read options from stdin using [_drush_backend_get_stdin()](https://github.com/drush-ops/drush/blob/8.x/includes/bootstrap.inc#L566).

**Problem**
In Windows 7 it's work good, `_drush_backend_get_stdin()` return array, but Windows 10 function return FALSE. This is because [stream_select($read, $write, $except, 1)](https://github.com/drush-ops/drush/blob/8.x/includes/backend.inc#L1276) always return `0`.

If comment (or delete) next code, function as expected will return array:

~~~
  if (drush_is_windows()) {
    $read = array($fp);
    $write = NULL;
    $except = NULL;
    $changed_streams = stream_select($read, $write, $except, 0);
    if ($changed_streams < 1) {
      return FALSE;
    }
  }
~~~

I don't know why `stream_select()` return `0`.

Drush version: 8.x-dev
PHP version: 7.2, 7.1, 7.0, 5.6 (tried at al)
Windows version: 10 (1809)
Shell: cmd, cygwin, git bash (tried at al)

**Comment by greg-1-anderson** [comment]
`stream_select()` is checking to see whether there is anything ready to be read on the stdin stream. The way this code is written presents a possible race condition. You could try adding a very short sleep before selecting, to see if that gives Win10 time enough to present stdin to php.

I don't know if there's a better way. In Drush 9 we no longer use backend invoke; we will remove it in Drush 10.

**Comment by xandeadx** [comment]
I try add `sleep(1)` before `stream_select()` and set stream_select `tv_set` param to `1` but it's does not help.

Without next code all works good:

~~~
    if ($changed_streams < 1) {
      return FALSE;
    }
~~~

**Comment by greg-1-anderson** [comment]
Maybe conditionalize the whole branch so that the stream_select is skipped on Windows 10 and later, but runs for Windows 7 and non-Windows machines. (I do not know what to advise for Windows 8.)

Note that if you do this, you might introduce a defect where `drush status --backend` will hang. In general, `--backend` is only used when stdin is redirected, so it should be rare to encounter this, and in any event this would be preferable to `drush make` not working at all.  It's also possible that Windows 10 might just work even without the Windows workaround, since this version is much closer to a *nix system. Try running any Drush command with `--backend` without redirecting stdin and see what happens.

**Comment by xandeadx** [comment]
You mean run `drush status --backend`? Shell await read or something like that (nothing happens):

![drush](https://user-images.githubusercontent.com/845389/52751153-840b2680-2fff-11e9-936f-7ddbbc59647a.gif)

**Comment by greg-1-anderson** [comment]
Yeah, that's the bug I was referring to. On a Linux system, `drush status --backend` will run, outputting a lot of json data.

**Comment by greg-1-anderson** [comment]
If you run `echo foo | drush status --backend`, then it will probably work.

**Comment by xandeadx** [comment]
The situation is as follows:

**Windows 10**
`drush status --backend` = **shell await read**
commented `if ($changed_streams < 1)` + `drush status --backend` = **shell await read**
commented `if (drush_is_windows())` + `drush status --backend` = **shell await read**
`echo foo | drush status --backend` = **output json**
commented `if ($changed_streams < 1)` + `echo foo | drush status --backend` = **output json**
commented `if (drush_is_windows())` + `echo foo | drush status --backend` = **output json**

**Windows 7**
`drush status --backend` = **output json**
commented `if ($changed_streams < 1)` + `drush status --backend` = **shell await read**
commented `if (drush_is_windows())` + `drush status --backend` = **shell await read**
`echo foo | drush status --backend` = **output json**
commented `if ($changed_streams < 1)` + `echo foo | drush status --backend` = **output json**
commented `if (drush_is_windows())` + `echo foo | drush status --backend` = **output json**

**Comment by xandeadx** [comment]
Updated:
On Windows 10 + PHP 5.6 `drush status --backend` output json, but `stream_select()` still return `0` and `drush make` not work.

**Comment by greg-1-anderson** [comment]
Yeah, if we converted `(drush_is_windows())` to `(drush_is_windows() && not_win10())`, then Drush would hang if --backend is used without providing stdin (unusual use case).  Not desirable, but better than we have, so I'd accept that as a workaround until someone figures out how to do the equivalent  stream_select operation.

Not sure how to detect win10.
